docs(frontend): showcase overflow detection (#831)
zama-ai/concrete: docs/execution-analysis/simulation.md

# Simulation

> During development, the speed of homomorphic execution can be a blocker for fast prototyping. You could call the function you're trying to compile directly, of course, but it won't be exactly the same as FHE execution, which has a certain probability of error (see [Exactness](../core-features/table_lookups.md#table-lookup-exactness)).

Reviewer: This doc
```suggestion
This document explains how to use the simulation feature to speedup prototyping and detect overflows during Fully Homomorphic Encryption (FHE) computation. 

Simulation allows to mimic the behavior of FHE execution without actually performing the encrypted computations.
```

Reviewer: This original sentence is not very clear to me:
" You could call the function you're trying to compile directly, of course, but **it** won't be exactly the same as FHE execution, **which** has a certain probability of error "

- Is this sentence describing the simulation feature? or the issue to be addressed by the simulation feature?
- What does "it" refers to - the simulated function?
- What doesn "which" refers to - the FHE function? of the simulation process?

I would suggest rewriting this part with better clarity. 
For example, 

> Simulations allow to.... (what does it do)  It is helpful to .....(why is it useful)

{% hint style="warning" %}

> There are some operations which are not supported in simulation yet. They will result in compilation failures. You can revert to simulation using graph execution using `circuit.graph(...)` instead of `circuit.simulate(...)`, which won't simulate FHE, but it will evaluate the computation graph, which is like simulating the operations without any errors due to FHE.

Reviewer: ```suggestion
Some operations are not yet supported in simulation and will result in compilation failures. In such cases, you can use graph execution with `circuit.graph(...)` instead of `circuit.simulate(...)`. This will evaluate the computation graph without simulating FHE errors.
```

{% endhint %}

## Overflow Detection in Simulation

> Overflow can happen during an FHE computation, leading sometimes to weird behaviors. Using simulation can help you detect such event. It prints a warning whenever an overflow happens. This feature is disabled by default, but you can enable it by using `detect_overflow_in_simulation=True` during compilation.

Reviewer: ```suggestion
Overflow can happen during an FHE computation, leading to unexpected behaviors. Using simulation can help you detect these events by printing a warning whenever an overflow happens. This feature is disabled by default, but you can enable it by setting `detect_overflow_in_simulation=True` during compilation.
```

> We can use the above circuit to trigger some overflow, but we will have to compile our circuit again to enable this feature

Reviewer: ```suggestion
To demonstrate, we will compile the previous circuit with overflow detection enabled and trigger an overflow:
```

```python
# compile with overflow detection enabled
circuit = f.compile(inputset, p_error=0.1, fhe_simulation=True, detect_overflow_in_simulation=True)
# cause an overflow
circuit.simulate([0,1,2,3,4,5,6,7,8,15])
```

> You should see the following warning after the simulation call

Reviewer: ```suggestion
You will see the following warning after the simulation call:
```

```bash
WARNING at loc("script.py":3:0): overflow happened during addition in simulation
```

> If you look at the MLIR (`circuit.mlir`), you will see that the input is supposed to be of type `eint4` thus represented in 4 bits (15 maximum). Since there's an addition of the input, we used the maximum value (15) here to trigger an overflow (15 + 1 = 16 which needs 5 bits). The warning tells us about the operation that did the overflow, as well as its location. You will get similar warnings for all basic FHE operations (add, mul, and lookup tables).

Reviewer: ```suggestion
If you look at the MLIR (`circuit.mlir`), you will see that the input type is supposed to be `eint4` represented in 4 bits with a maximum value of 15. Since there's an addition of the input, we used the maximum value (15) here to trigger an overflow (15 + 1 = 16 which needs 5 bits). The warning specifies the operation that caused the overflow and its location. Similar warnings will be displayed for all basic FHE operations such as add, mul, and lookup tables.
```
